x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/autoscaling/AutoscalerService.java

```diff
@@ -0,0 +1,218 @@
+        this.client = client;
+        this.isNlpEnabled = isNlpEnabled;
+
+        lastInferenceStatsByDeploymentNode = new HashMap<>();
+        autoscalers = new HashMap<>();
+    }
+
+    public void start() {
+        update(clusterService.state());
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        update(event.state());
+    }
+
+    private synchronized void update(ClusterState state) {
+        if (isNlpEnabled == false) {
+            return;
+        }
+
+        TrainedModelAssignmentMetadata assignments = TrainedModelAssignmentMetadata.fromState(state);
+        for (TrainedModelAssignment assignment : assignments.allAssignments().values()) {
+            if (assignment.getAutoscalingSettings() != null && assignment.getAutoscalingSettings().getEnabled()) {
+                Autoscaler autoscaler = autoscalers.computeIfAbsent(
+                    assignment.getDeploymentId(),
+                    key -> new Autoscaler(assignment.getDeploymentId(), assignment.totalTargetAllocations())
+                );
+                autoscaler.setMinMaxNumberOfAllocations(
+                    assignment.getAutoscalingSettings().getMinNumberOfAllocations(),
+                    assignment.getAutoscalingSettings().getMaxNumberOfAllocations()
+                );
+            } else {
+                autoscalers.remove(assignment.getDeploymentId());
+            }
+        }
+
+        if (autoscalers.isEmpty() == false) {
+            if (cancellable == null) {
+                logger.debug("Starting ML inference autoscaler");
+                scheduleNextTrigger();
+            }
+        } else {
+            stop();
+        }
+    }
+
+    public synchronized void stop() {
+        if (cancellable != null && cancellable.isCancelled() == false) {
+            logger.debug("Stopping ML inference autoscaler");
+            cancellable.cancel();
+            cancellable = null;
+        }
+    }
+
+    private void scheduleNextTrigger() {
+        try {
+            cancellable = threadPool.schedule(this::trigger, TimeValue.timeValueSeconds(TIME_INTERVAL_SECONDS), threadPool.generic());
+        } catch (EsRejectedExecutionException e) {
+            if (e.isExecutorShutdown() == false) {
+                throw e;
+            }
+        }
+    }
+
+    private synchronized void trigger() {
+        scheduleNextTrigger();
+        getDeploymentStats(ActionListener.wrap(this::processDeploymentStats, e -> logger.warn("Error in inference autoscaling", e)));
+    }
+
+    private void getDeploymentStats(ActionListener<GetDeploymentStatsAction.Response> processDeploymentStats) {
+        String deploymentIds = String.join(",", autoscalers.keySet());
+        ClientHelper.executeAsyncWithOrigin(
+            client,
+            ClientHelper.ML_ORIGIN,
+            GetDeploymentStatsAction.INSTANCE,
+            new GetDeploymentStatsAction.Request(deploymentIds),
```

Comment:
I mentioned before that `GetDeploymentStatsAction` does quite a lot of work, we should add an option to the request to return the subset of the stats required for the autoscaling calculation. I'll take that change

Reply:
Cool, thanks. I've added a TODO for that.